### PR TITLE
Prefer SHGetKnownFolderPath() in g_get_home_dir() over HOMEDRIVE/HOMEPATH environment variables

### PR DESCRIPTION
**Pull request by ntherning**
Relying on the HOMEDRIVE/HOMEPATH environment variables is fragile as those variables don't always have to be set. One such scenario is when running on a Jenkins slave. When not set g_get_home_dir() in eglib would return NULL which led to a crash in ves_icall_System_Environment_InternalGetHome().

This patch changes g_get_home_dir() to first try to call SHGetKnownFolderPath() to get the current user's profile directory (i.e. home dir). If that fails it falls back to the USERPROFILE environment variable. If that isn't available the function will fallback to the original behavior and try with HOMEDRIVE/HOMEPATH.

### CODE REVIEW

**Comment by ntherning** [comment]
build

File: eglib/src/gmisc-win32.c [hunk 1]
```diff
@@ -116,20 +117,32 @@ g_path_is_absolute (const char *filename)
...
 	gchar *home_dir = NULL;
-	
-	if (drive && path) {
-		home_dir = g_malloc(strlen(drive) + strlen(path) +1);
-		if (home_dir) {
-			sprintf(home_dir, "%s%s", drive, path);
-		}
+
+	PWSTR profile_path = NULL;
+	HRESULT hr = SHGetKnownFolderPath (&FOLDERID_Profile, KF_FLAG_DEFAULT, NULL, &profile_path);
```
Comment (akoeplinger) [inline eglib/src/gmisc-win32.c]: Looks like mingw doesn't know SHGetKnownFolderPath (or we somehow end up not defining _WIN32_WINNT=0x0600 in eglib):

```
gmisc-win32.c: In function 'monoeg_g_get_home_dir':
gmisc-win32.c:123:2: warning: implicit declaration of function 'SHGetKnownFolderPath' [-Wimplicit-function-declaration]
  HRESULT hr = SHGetKnownFolderPath (&FOLDERID_Profile, KF_FLAG_DEFAULT, NULL, &profile_path);
  ^
gmisc-win32.c:123:56: error: 'KF_FLAG_DEFAULT' undeclared (first use in this function)
  HRESULT hr = SHGetKnownFolderPath (&FOLDERID_Profile, KF_FLAG_DEFAULT, NULL, &profile_path);
```

What we've done in the past is simply including the declaration ourselves wrapped in `#if !defined`

Comment (ntherning) [inline eglib/src/gmisc-win32.c]: Thanks for taking a look! You are right, there seems to be a bug in the `configure.ac` in the root. It never passes on any defines at all to `eglib/configure`, like it does when invoking `libgc/configure`. Mingw does have the required header file and knows about `SHGetKnownFolderPath()` but only when `_WIN32_WINNT` is set to at least Vista. I'll see if I can sort that out too...

**Comment by ntherning** [comment]
Updated the PR to only try with `SHGetKnownFolderPath()` when compiled with VS. Couldn't get mingw to link against shell32 and uuid properly I'm afraid.

**Comment by akoeplinger** [comment]
LGTM.

**Pull request merged**